sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -883,6 +883,28 @@ createKernelArgMask(const pi::ByteArray &Bytes) {
   return Result;
 }
 
+static void handleXilinxKernels(RTDeviceBinaryImage* Img) {
+  static bool has_been_invoked = false;
+  const char* target = Img->getTarget();
+  const char* env_var = "XCL_EMULATION_MODE";
+  if (strncmp("fpga64_", target, 7) == 0 || strncmp("fpga32", target, 7) == 0) {
+    target += 7;
+    if (strcmp(target, "hw") == 0) {
+      if (has_been_invoked)
+        assert(!std::getenv(env_var));
```

Comment:
I have the feeling that all these asserts should actually be replaced by some SYCL exception `throw`.
Because the runtime could be compiled with `-DNDEBUG` to remove the `assert()` macros leaving a lot of undefined behavior all over the place.

Reply:
we could throw be this code is run duding startup of the executable before main so the user as no chance of catching that throw.